### ISSUE TITLE
[BUG] Misalignment around silences

### ISSUE THREAD

**Issue by DanielSWolf** [MontrealCorpusTools/Montreal-Forced-Aligner#386]
**Debugging checklist**

[x] Have you updated to latest MFA version?
[x] Have you tried rerunning the command with the `--clean` flag?

**Describe the issue**

I've successfully trained my own acoustic model, then I used `mfa align` to align some recordings. For the most part, the results are excellent, including any pauses. However, for one file with leading silence, the first word is aligned to start at 0.0s, even if the signal clearly doesn't start until later.

See this screenshot of Praat. The first word, "ho", is aligned from 0.0s to 0.16s, even though visually, it's apparent that the speech doesn't start until 0.2s. The rest of the alignment is perfect, including the very next word, "i".

![image](https://user-images.githubusercontent.com/5963162/148601441-9ea704b6-6c56-4874-94e4-3f347bcc8e6b.png)

**For Reproducing your issue**
Please fill out the following:

1. Corpus structure
   * What language is the corpus in? Italian
   * How many files/speakers? Just a handful of files to test the alignment.
   * Are you using lab files or TextGrid files for input? lab.
2. Dictionary
   * Are you using a dictionary from MFA? If so, which one? No, a custom Italian dictionary.
   * If it's a custom dictionary, what is the phoneset? It's IPA, but with non-trivial phonemes like "mː" and "lː" that MFA doesn't know about. So I didn't specify a phone set during training.
3. Acoustic model
   * If you're using an acoustic model, is it one download through MFA? If so, which one? I trained my own Italian acoustic model.
   * If it's a model you've trained, what data was it trained on? It was trained on the Italian MLS (Multilingual LibriSpeech) corpus: 62,133 files by 85 speakers.

**Log file**

Please attach the log file for the run that encountered an error (by default these will be stored in `~/Documents/MFA`). There was no error, just a misalignment. I'm attaching the log file anyway: [pretrained_aligner.log](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner/files/7831201/pretrained_aligner.log)

**Desktop (please complete the following information):**
 - OS: Linux
 - Version: Ubuntu 20.04.3 LTS
 - Any other details about the setup: I'm using the official Docker image for MFA 2.0.0b11.

**Comment by DanielSWolf** [comment]
On further inspection, this problem doesn't seem to be limited to the beginning of a file. It seems that if a file contains silences between utterances, sometimes the word following the silence is "stretched back" so it fills the silence, like "il" in this example (which should be in the red range):

![image](https://user-images.githubusercontent.com/5963162/148652192-3e950c39-257b-4336-b6af-c521ddfe1fb5.png)

... while other times, the word following the silence is "shifted back" before the silence, like "e" in this example (which again should be in the red range):

![image](https://user-images.githubusercontent.com/5963162/148652151-0293aac7-2ee1-4061-9e53-916092d2d723.png)

Is there an option that tells MFA to expect silences before and after each word? I searched the documentation, but couldn't find anything along those lines.

**Comment by mmcauliffe** [comment]
This is definitely a weird one.  I tried running some English alignments with 2.0.0b11, and they aren't showing the same behavior, so I don't think it's a general bug, but could you try running the alingment with the `--disable_textgrid_cleanup`? That should stop MFA trying to clean up silences, so we'd be able to see the underlying silence phones and see if it's a problem with the acoustic model or with the textgrid cleanup.

**Comment by DanielSWolf** [comment]
Thank you for the suggestion! I re-ran the alignment with `--disable_textgrid_cleanup`. The `sil` placeholders are exactly where the gaps in the timeline used to be:

![image](https://user-images.githubusercontent.com/5963162/148678635-ebccaff9-d463-4b5e-b7a2-06ed8cbd967d.png)
![image](https://user-images.githubusercontent.com/5963162/148678667-e29c2428-2fe0-41bb-8800-b2e35c0b7056.png)

So the problem is *not* with the new cleanup logic for silences, but with the alignment itself.

I'm attaching a zip file containing the acoustic model, the dictionary, the audio file and its transcript. 
[alignment-test.zip](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner/files/7834882/alignment-test.zip)

**Note:** When I align only this single file (as opposed to aligning it as part of a larger corpus), the misalignment in the middle doesn't occur, only the one at the start. 🤔

Please let me know if there is anything else I can do to help!

**Comment by mmcauliffe** [comment]
So I looked at the model and its transitions and it looks like it's doing
the "right" thing for it, so I think something might have gone wrong while
training. I pushed a new version out on conda last night and it's been
working well for the model I've been training, so maybe try retraining and
see if the alignments make more sense?

If that doesn't work, I'll set up a training run myself and see what
happened.

**Comment by DanielSWolf** [comment]
Sounds good! I'll give it a try and get back to you!

**Comment by mmcauliffe** [comment]
Hmm, weird, I'll try to do a run myself this weekend and see what bugs pop out.  What's the exact command that you're using to train? Any differences from the default settings?

For your questions:

1. It is 0 that were too short to be aligned, previously it captured information about the cause of alignment failures, but in some refactors to how alignment is run and what gets sent back to the main thread, that error message no longer makes sense.  I'll try to move it earlier in the process, since the logs for MFCC should have that information sooner.
2. Utterance for lab-file format does indeed equal a file (TextGrid input format has more utterances than files).  The "too short" message initially corresponds to it not being able to be aligned with the given beam.
3. The utterance naming scheme is: `speaker_id`-`file_name`-`begin_time_stamp`-`end_time_stamp`.  So for `7458-6589-000135-0-17-32`, it's speaker `7458`, file `6589-000135`, begin `0`, end `17.32` (for lab-based, begin will always be 0, end will be the duration of the file).

**Comment by DanielSWolf** [comment]
Thanks for your answers -- that cleared up my confusion!

When you say you'll try it yourself: Do you mean with my exact data? In this case, I'd try and send them to you in a format that's quick to set up.

**Comment by mmcauliffe** [comment]
Didn't you use the Italian Librispeech as the training data? I was just gonna download, transform it and train.

**Comment by DanielSWolf** [comment]
Well, more or less! 😉 Follow these steps:

1. Download and extract the attached file [train.tar.gz](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner/files/7872559/train.tar.gz) (8 MB). You'll get a `train` directory containing all you need for training except for the actual audio files:
   * `train/it.dict`, an Italian dictionary trained from a custom G4P model that contains exactly the words used in the Italian MLS corpus
   * `train/corpus/`, a directory structure containing .lab files for the Italian MLS corpus that have been carefully normalized to follow the same conventions as the dictionary.
2. Download the [FLAC-compressed version of the Italian MLS corpus](https://dl.fbaipublicfiles.com/mls/mls_italian.tar.gz) (15 GB).
3. Run `tar -zxvf mls_italian.tar.gz --strip-components 5 -C <some target directory>`. This will remove the (undesired) directory structure and give you a flat list of FLAC files.
4. Convert all FLAC files to WAVE.
5. Using a tool of your choice, fix the structure and naming of the WAVE files: `<speaker>_<book>_<counter>.wav` must become `<speaker>/<book>_<counter>.wav`, e.g. `10446_10415_001569.wav` -> `10446/10415_001569.wav`.
6. Now the structure and naming of the WAVE files match that of the .lab files. Copy the WAVE files over to the `train` directory you created initially.
7. Run `mfa validate train/corpus/ train/it.dict` (no special flags or config).
8. Run `mfa train train/corpus/ train/it.dict train/acoustic-model.zip train/text-grids` (dito).

**Comment by DanielSWolf** [comment]
I'm running the training script again just now, just to make sure the first run wasn't some kind of a fluke. It's still running, but looking at `~/Documents/MFA/corpus_train_acoustic_model/monophone_ali/alignment_log_likelihood.csv` and `~/Documents/MFA/corpus_train_acoustic_model/triphone_ali/alignment_log_likelihood.csv`, I can see that the vast majority of utterances has no log-likelihood at all, which -- I take it -- means training alignment is failing in most cases:

```text
utterance,loglikelihood
7458-6589-000124-0-18-1,
7458-6589-000127-0-12-47,
7458-6589-000103-0-11-73,
7458-6589-000037-0-15-36,
7458-6589-000035-0-12-03,
7458-6589-000000-0-10-25,
7458-6589-000004-0-15-18,
7458-6589-000038-0-15-29,-107.05506578947369
7458-6589-000017-0-13-28,
7458-6589-000018-0-17-07,
7458-6589-000048-0-12-69,
7458-6589-000013-0-10-33,
7458-6589-000083-0-10-3,
7458-6589-000095-0-14-82,
7458-6589-000077-0-15-92,
...
```

Thank you very much in advance for taking the time to look into this! I'm looking forward to hearing how your fared!

**Comment by mmcauliffe** [comment]
Ok, I think I have it figured out, running a training test right now, and should hopefully get a release out that fixes the bug later today.  If you want to get it started before then, try keeping the speaker IDs when moving files, MFA was getting confused about what file name pointed to what utterance.  The fix I'm testing should avoid that confusion.

**Comment by mmcauliffe** [comment]
Also there might be some data quality issues, speaker 166 seems to have both a male and a female speakers, don't know if that's true of yours and whether there are other such speakers, but that's going to be troublesome for the training (though not likely to garble it like the file confusion issue)

**Comment by DanielSWolf** [comment]
Thank you so much for looking into it!

I see how missing the first word in the audio (in other words, having one extra initial word in the transcript) might trick the model into learning that initial silence is really an arbitrary word in disguise. I guess I'll have to do some cleanup there. (I can imagine running an initial training/alignment round, running an extra VAD on the files, removing any files from the corpus whose first aligned word falls into the initial silence, and then running the final training on the remaining corpus.)

I also understand that having one speaker ID with multiple actual speakers may deteriorate the results.

Let me make sure I understand your point about "keeping the speaker IDs when moving files", though. My corpus structure looks like this:

```text
corpus
├── 10121
│   ├── 9630_000000.lab
│   ├── 9630_000000.wav
│   ├── 9630_000001.lab
│   ├── 9630_000001.wav
│   ├── 9630_000002.lab
│   ├── 9630_000002.wav
│   ...
├── 10179
│   ├── 9630_000000.lab
│   ├── 9630_000000.wav
│   ├── 9630_000001.lab
│   ├── 9630_000001.wav
│   ├── 9630_000002.lab
│   ├── 9630_000002.wav
│   └── ...
├── 10191
│   ├── 9630_000000.lab
│   ├── 9630_000000.wav
│   ├── 9630_000001.lab
│   ├── 9630_000001.wav
│   ├── 9630_000002.lab
│   ├── 9630_000002.wav
│   └── ...
└── ...
```

My corpus has one directory per speaker, and each speaker directory contains .wav and .lab files with matching names. Are you saying that these file names must be *globally* unique, that is, that no two speaker directories must contain audio files with identical names? If MFA relies on the audio file names being globally unique, this would certainly explain why only a fraction of the files got aligned.

**Comment by mmcauliffe** [comment]
Yeah, so one of the recent refactors used a dictionary to store files based on their file names, so if there were duplicate wav/lab file names (ignoring directory structure), then it would overwrite and lose track of earlier files.  I'm trying to track down one last transient bug unrelated to this (I think), and then I'll push a release out.  In the meantime, here's the model that looks to be doing fairly well with alignments: [italian.zip](https://github.com/MontrealCorpusTools/Montreal-Forced-Aligner/files/7877386/italian.zip).

**Comment by mmcauliffe** [comment]
Ok, version 2.0.0rc2 should be up on pypi/conda with the fixes for this if you want to try it, just be sure to include `--clean` to remove any lingering issues from the earlier runs.

**Comment by DanielSWolf** [comment]
Thank you very much!
